### PR TITLE
Feature/user authentication

### PR DESCRIPTION
**Pull request by sagg0t**

### CODE REVIEW

File: config/locales/devise.security_extension.ja.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+ja:
```
Comment (dmbavykin) [inline config/locales/devise.security_extension.ja.yml:1]: 👍

File: app/controllers/books_controller.rb [hunk 2]
```diff
@@ -4,12 +4,12 @@ class BooksController < ApplicationController
 
   def index
     @category_book_count = CategoryCountQuery.call
-    @categories = Category.all
+    # @categories = Category.all
```
Comment (dmbavykin) [inline app/controllers/books_controller.rb]: 😨 ?

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  def on_login_success
+    sign_in_and_redirect @user, event: :authentication
+
+    return unless is_navigational_format?
+
+    set_flash_message :notice, :success, kind: 'Facebook'
+  end
+
+  def on_login_failure
+    session['devise.facebook_data'] = request.env['omniauth.auth']
```
Comment (dmbavykin) [inline app/controllers/users/omniauth_callbacks_controller.rb]: why r u store it in session?

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  def self.from_omniauth(auth)
+    where(email: auth.info.email).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[8..20]
+      user.provider = auth.provider
+      user.uid = auth.uid
+    end
+  end
+
+  validates :email, uniqueness: true
```
Comment (dmbavykin) [inline app/models/user.rb]: move it higher than method definition

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+Devise.setup do |config|
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (dmbavykin) [inline config/initializers/devise.rb]: r u think that ur mailer works correctly? :)

File: spec/features/authentication/sign_in_spec.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
+describe 'Sign In', type: :feature do
```
Comment (dmbavykin) [inline spec/features/authentication/sign_in_spec.rb]: It`s unnecessary to define type of spec with [this](https://relishapp.com/rspec/rspec-rails/docs/directory-structure) configuration

File: spec/queries/book_sort_query_spec.rb [hunk 7]
```diff
@@ -1,7 +1,7 @@
 describe BookSortQuery do
-  let(:book_a) { create(:book, title: 'a', price: 1, created_at: Date.current - 3.day) }
-  let(:book_b) { create(:book, title: 'b', price: 2, created_at: Date.current - 2.day) }
-  let(:book_c) { create(:book, title: 'c', price: 3, created_at: Date.current - 1.day) }
+  let(:book_a) { create(:book, title: 'a', price: 1, created_at: 3.days.ago) }
```
Comment (dmbavykin) [inline spec/queries/book_sort_query_spec.rb:2]: 👍

File: spec/support/capybara.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# Capybara.register_driver :selenium do |app|
+#   Capybara::Selenium::Driver.new(app, browser: :chrome)
+# end
+
+Capybara.configure do |config|
```
Comment (dmbavykin) [inline spec/support/capybara.rb]: https://thoughtbot.com/blog/headless-feature-specs-with-chrome
can u try to use headless chrome instead?

**Review by dmbavykin — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 9]
```diff
@@ -43,9 +42,11 @@ jobs:
 
       - run:
           name: '***** DB SETUP *****'
-          command: |
-            bundle exec rails db:create
-            bundle exec rails db:schema:load
+          command: bundle exec rails db:setup
+
+      - run:
```
Comment (dmbavykin) [inline .circleci/config.yml:47]: Screenshot 2019-12-15 at 10 50 42
what about other linters? 🤔

**Review by dmbavykin — CHANGES_REQUESTED** [review]

File: app/mailers/application_mailer.rb [hunk 10]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: 'no-reply@bookstore-rg-ov.herokuapp.com'
```
Comment (Svatok) [inline app/mailers/application_mailer.rb]: Move to constant

File: config/environments/development.rb [hunk 11]
```diff
@@ -17,8 +17,24 @@
...
 
   config.active_storage.service = :local
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
   config.action_mailer.raise_delivery_errors = false
   config.action_mailer.perform_caching = false
+  # config.action_mailer.delivery_method = :letter_opener
+  # config.action_mailer.default_url_options = { host: 'bookstore-rg-ov.herokuapp.com' }
+  config.action_mailer.perform_caching = false
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
```
Comment (Svatok) [inline config/environments/development.rb]: remove this

File: config/environments/production.rb [hunk 12]
```diff
@@ -63,7 +63,18 @@
   # config.active_job.queue_adapter     = :resque
   # config.active_job.queue_name_prefix = "bookstore_#{Rails.env}"
 
+  config.action_mailer.default_url_options = { host: 'bookstore-rg-ov.herokuapp.com' }
```
Comment (Svatok) [inline config/environments/production.rb]: to constant

File: spec/features/authentication/sign_up_spec.rb [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        fill_in 'user[email]', with: email
+        fill_in 'user[password]', with: password[:allowed]
+        fill_in 'user[password_confirmation]', with: password[:allowed]
+      end
+
+      click_button I18n.t('user_actions.sign_up')
+
+      expect(page).to have_current_path root_path
+      find_link(I18n.t('pages_titles.account'), match: :first).click
+      expect(page).to have_content I18n.t('user_actions.sign_out')
```
Comment (Svatok) [inline spec/features/authentication/sign_up_spec.rb:30]: Add tests with email opening - check email body. gem 'capybara-email'

File: spec/features/book_page_spec.rb [hunk 14]
```diff
@@ -1,8 +1,9 @@
 describe 'Book page' do
-  let(:book) { create(:book, authors: [create(:author)]).decorate }
-
   before { visit book_path(id: book.id) }
 
+  let!(:book) { create(:book, authors: [create(:author)]).decorate }
```
Comment (Svatok) [inline spec/features/book_page_spec.rb]: Add trait to book factory with name `:with_authors`

**Review by Svatok — COMMENTED** [review]
http://rspec.info/blog/2016/07/rspec-3-5-has-been-released/

> For new Rails apps: we don't recommend adding the rails-controller-testing gem to your application. The official recommendation of the Rails team and the RSpec core team is to write request specs instead. Request specs allow you to focus on a single controller action, but unlike controller tests involve the router, the middleware stack, and both rack requests and responses. This adds realism to the test that you are writing, and helps avoid many of the issues that are common in controller specs. In Rails 5, request specs are significantly faster than either request or controller specs were in rails 4, thanks to the work by Eileen Uchitelle1 of the Rails Committer Team.

**Review by Svatok — APPROVED** [review]

**Pull request merged**